File: backend/internal/sales/service/manual_sale.go
package service

import (
	"context"
	"time"

	"github.com/peter/ticket_pos/backend/internal/platform"
	"github.com/peter/ticket_pos/backend/internal/sales/importfile"
	"github.com/peter/ticket_pos/backend/internal/sales/repository"
)

// RecordedManualSale is what a Manually Recorded Sale hands back: the sale, its
// Sale Confirmation reference, and enough of the sale itself for the modal's
// session receipt to show a line per save without a second round trip.
//
// It carries the duplicate warning too. A duplicate never blocks the record
// (ADR 0052) — two people at the door with the same name on the same night is a
// thing that happens — but the organizer twenty names into a notebook is
// exactly the person who wants to be told they may just have typed one twice,
// and the receipt is where they will see it.
//
// ConfirmationSent is reported rather than assumed. The buyer is ALWAYS mailed,
// with no toggle and no request field, so this is false only when the mail
// itself failed — which does not undo the sale, and which the caller is
// entitled to know about.
type RecordedManualSale struct {
	SaleID            string    `json:"sale_id"`
	ConfirmationRef   string    `json:"confirmation_ref"`
	CustomerEmail     string    `json:"customer_email"`
	CustomerFirstName string    `json:"customer_first_name"`
	CustomerLastName  string    `json:"customer_last_name"`
	TicketTypeID      string    `json:"ticket_type_id"`
	TicketTypeName    string    `json:"ticket_type_name"`
	Quantity          int       `json:"quantity"`
	AmountCents       int       `json:"amount_cents"`
	Currency          string    `json:"currency"`
	SoldAt            time.Time `json:"sold_at"`
	PaymentMethod     string    `json:"payment_method"`
	ConfirmationSent  bool      `json:"confirmation_sent"`
	PossibleDuplicate bool      `json:"possible_duplicate"`
	DuplicateOfDate   string    `json:"duplicate_of_date,omitempty"`
}

// RecordManualSale records ONE Manually Recorded Sale (#368, parent #366,
// ADR 0052): a Sale Import row typed into a form instead of uploaded in a file.
//
// IT IS AN IMPORT ROW IN EVERY RESPECT BUT THE SPREADSHEET. The `import` Sales
// Channel, Sales Source `direct`, one Ticket Type, a Payment Method, a Sale
// Confirmation to the buyer, Tickets minted one per unit and capacity taken —
// and it is judged by the very validator the Sale Import preview and file commit
// use, through the shared typed-row verdict (#367). That sharing is not a
// convenience: two routes now record the same object, and if they validated
// separately the same act would succeed one way and fail the other for reasons
// nobody could see.
//
// IT IS NOT A SALE IMPORT. It belongs to no batch, so it never appears in the
// Import history and no batch undo can reach it — and, the point that decided
// the design, recording one never makes an earlier upload stop being the latest
// batch, so an existing batch stays undoable. Minting a one-row batch per typed
// sale would have quietly disarmed that guard.
//
// THE BUYER IS ALWAYS MAILED, with no toggle and no request field, as a Sale
// Import does and unlike a Sale Correction. ADR 0050 made the correction's
// Confirmation off-by-default because the buyer already holds one from the
// original import; no prior mail exists here, and suppressing it would ship a
// sale whose buyer holds no Confirmation Link and therefore cannot assign a
// Ticket or answer a Ticket Question. A mail that fails does not undo the sale:
// the sale is the record of money that changed hands, and the mail is a
// notification about it.
//
// THERE IS NO IDEMPOTENCY KEY, deliberately — there is no batch to hang one on,
// and Sale Correction, the closest precedent, has none either. The accepted hole
// and its mitigations are in ADR 0052's consequences.
//
// A refused row returns the verdict with a nil result and NOTHING WRITTEN: the
// caller reports one complaint per offending column. The transaction's own
// capacity check under lock remains the last word on a race lost after the
// verdict was reached, and surfaces as IMPORT_BATCH_FAILED.
func (s *Service) RecordManualSale(ctx context.Context, actor ActorContext, eventID string, in ImportRowInput) (*RecordedManualSale, *importfile.ValidateResult, error) {
	// An externally registered Event is refused here, before the body is judged
	// — loadImportContext's own guard, and the reason it runs before the Ticket
	// Types are read (ADR 0028). Such an Event has no Ticket Types, so every
	// body names one that does not exist, and "unknown ticket type" would send
	// the organizer hunting for a data problem that is not there.
	event, types, loc, err := s.loadImportContext(ctx, actor.OrganizationID, eventID)
	if err != nil {
		return nil, nil, err
	}

	validated, err := s.validateManualSale(ctx, actor.OrganizationID, eventID, types, loc, in)
	if err != nil {
		return nil, nil, err
	}
	if !validated.Valid() {
		return nil, validated, nil
	}
	row := validated.Rows[0]

	ref, err := generateConfirmationRef()
	if err != nil {
		return nil, nil, err
	}
	recorded, err := s.repo.RecordBatchlessImportSale(ctx, repository.RecordBatchlessImportSaleInput{
		EventID:        eventID,
		OrganizationID: actor.OrganizationID,
		Now:            s.now(),
		UpsertCustomer: s.customers.UpsertForSale,
		// The buyer holds Ticket 1 (ADR 0055). A typed row is a transcription
		// exactly as a file row is — the Member is recording a sale that already
		// happened — and there is deliberately no checkbox to opt out of it:
		// a Holder differing from the buyer must be written `assigned`, which
		// ADR 0047 shows no name for.
		SelfHeld: s.ticketAssignmentEnabled,
		Sale: repository.CommitSale{
			// No phone and no self-assertion, exactly as a file import records
			// none: an Organization's account of a sale it took off-platform
			// never proves the buyer owns the email. The Customer is upserted
			// unverified for the same reason.
			Customer: platform.SaleCustomer{
				Email:     row.CustomerEmail,
				FirstName: row.CustomerFirstName,
				LastName:  row.CustomerLastName,
				TaxID:     platform.SaleTaxID{Type: row.CustomerTaxIDType, Number: row.CustomerTaxIDNumber},
			},
			PaymentMethod:   row.PaymentMethod,
			SoldAt:          row.SoldAtTime(),
			ConfirmationRef: ref,
			// ONE LINE, ALWAYS. No `import`-channel sale has ever carried more
			// than one Ticket Sale Line and this route does not introduce the
			// first: the family buying two Adult and two Child records two
			// sales, which ADR 0052 acknowledges as a cost rather than denying.
			// A nil UnitPriceCents snapshots the Ticket Type's catalog price,
			// as a blank amount cell does; zero records a comp.
			Lines: []repository.CommitLine{{
				TicketTypeID:   row.TicketTypeID,
				Quantity:       row.Quantity,
				UnitPriceCents: row.AmountCents,
			}},
		},
	})
	if err != nil {
		return nil, nil, mapCommitError(err)
	}

	sent := s.email.SendSaleConfirmation(ctx, s.importedSaleConfirmation(ctx, event, *recorded)) == nil

	return &RecordedManualSale{
		SaleID:            recorded.ID,
		ConfirmationRef:   recorded.ConfirmationRef,
		CustomerEmail:     recorded.CustomerEmail,
		CustomerFirstName: recorded.CustomerFirstName,
		CustomerLastName:  recorded.CustomerLastName,
		TicketTypeID:      row.TicketTypeID,
		TicketTypeName:    row.TicketTypeName,
		Quantity:          row.Quantity,
		AmountCents:       recorded.AmountCents,
		Currency:          event.Currency,
		SoldAt:            row.SoldAtTime(),
		PaymentMethod:     row.PaymentMethod,
		ConfirmationSent:  sent,
		PossibleDuplicate: row.PossibleDuplicate,
		DuplicateOfDate:   row.DuplicateOfDate,
	}, nil, nil
}

// PreviewManualSale is the manual form's live verdict (#368): the row judged
// exactly as RecordManualSale would judge it, in the Sale Import preview's own
// shape, with nothing written.
//
// It exists twice over. It is the surface that tells the organizer what is
// wrong in the field that caused it, as they type — and it is the per-field
// complaint channel whose absence is the JSON commit's stated reason for
// exempting itself from the Purchase Limit. Supplying one is what makes
// enforcing the limit here reasonable rather than merely stricter (ADR 0052).
func (s *Service) PreviewManualSale(ctx context.Context, actor ActorContext, eventID string, in ImportRowInput) (*importfile.ValidateResult, error) {
	_, types, loc, err := s.loadImportContext(ctx, actor.OrganizationID, eventID)
	if err != nil {
		return nil, err
	}
	return s.validateManualSale(ctx, actor.OrganizationID, eventID, types, loc, in)
}

// validateManualSale judges the typed row as the shared verdict judges any
// single one (#367), and is the whole of both the preview and the gate the
// record passes through — so the live verdict and the refusal can never
// disagree.
//
// It supplies the ZERO typedRowContext, and both halves of that are the
// statement: nothing is being reversed, so no Ticket Sale is left out of the
// duplicate signal or the Purchase Limit tally — a sale recorded earlier in the
// same sitting is a real sale and counts against the limit like any other — and
// the capacity complaint is the plain shortfall, because the catalog snapshot it
// is measured against is simply the Event's, with nothing netted back into it.
// Its sibling, a Sale Correction, does the opposite on both counts.
//
// The exclusion for the Purchase Limit is "" for the same reason the file import
// passes "": there is no prior sale in the picture at all.
func (s *Service) validateManualSale(ctx context.Context, orgID, eventID string, types []importfile.TicketTypeRef, loc *time.Location, in ImportRowInput) (*importfile.ValidateResult, error) {
	return s.validateTypedImportRow(ctx, orgID, eventID, types, loc, in, typedRowContext{})
}

File: apps/staff/lib/manual-sale.ts
/**
 * The record-a-sale form's state, body and verdict (#369, ADR 0052).
 *
 * A Manually Recorded Sale is one Sale Import row typed instead of uploaded, so
 * everything here is about ONE row: what the inputs hold, what body the save
 * sends, and how the preview's answer is read. It is a separate module from
 * sales-api.ts because that one is the Sales list's endpoints and tokens, and
 * this is a form's state machine with no fetch in it.
 *
 * TOKENS AND STATE, NEVER SENTENCES. Nothing here imports a catalog, React or
 * next-intl, which is what lets manual-sale.test.ts assert a decision under
 * `node --test` — the staff app has no component tests and adds none, so a rule
 * that is not a pure function here is a rule nothing defends.
 */

// Relative (not "@/lib") so the module graph resolves under `node --test` as
// well as the bundler — the unit tests import this file directly.
import { dateTimeLocalToISO } from "./events-api.ts";
import type { ImportPreviewResult } from "./imports-api.ts";
import type { RecordSaleInput } from "./sales-api.ts";

/**
 * The form as typed: every template column a string, because that is what an
 * input holds. The sale date is a datetime-local value in the EVENT's zone and
 * the amount is a price per ticket in major units; both are converted on the way
 * out and never before, so nothing the organizer typed is rewritten under them.
 */
export type ManualSaleForm = {
  customerEmail: string;
  customerFirstName: string;
  customerLastName: string;
  taxIdType: string;
  taxIdNumber: string;
  ticketTypeId: string;
  quantity: string;
  paymentMethod: string;
  soldAtLocal: string;
  unitPrice: string;
};

/**
 * The template column each field's complaints arrive under.
 *
 * The API names BARE columns — no `rows[N].` prefix, the convention for a single
 * typed row — and the form files each complaint under the input that caused it.
 * `ticketTypeId` is the one asymmetry: the body sends `ticket_type_id` and a
 * refusal blames `ticket_type`, the name the uploaded template prints.
 */
export const MANUAL_SALE_COLUMNS = {
  customerEmail: "customer_email",
  customerFirstName: "customer_first_name",
  customerLastName: "customer_last_name",
  taxIdType: "customer_tax_id_type",
  taxIdNumber: "customer_tax_id_number",
  ticketTypeId: "ticket_type",
  quantity: "quantity",
  paymentMethod: "payment_method",
  soldAtLocal: "sold_at",
  unitPrice: "amount",
} as const satisfies Record<keyof ManualSaleForm, string>;

/**
 * A blank form on the given sale date, which the caller draws as today on the
 * EVENT's clock — recording the sale just taken should need no thought, and the
 * reader's laptop is never asked what day it is.
 *
 * Quantity starts at one, the sale a person types by hand. The amount starts
 * blank, which means the Ticket Type's own price: the safe default and the
 * common case.
 */
export function emptyManualSaleForm(soldAtLocal: string): ManualSaleForm {
  return {
    customerEmail: "",
    customerFirstName: "",
    customerLastName: "",
    taxIdType: "",
    taxIdNumber: "",
    ticketTypeId: "",
    quantity: "1",
    paymentMethod: "",
    soldAtLocal,
    unitPrice: "",
  };
}

/**
 * manualSaleQuantity reads the quantity cell into the number the body carries.
 *
 * It does not round and it does not judge. A fractional quantity travels as
 * typed, because the API answers 1.5 on the quantity column in the import
 * validator's own words, and rounding it here would record a sale nobody typed.
 * Anything unreadable travels as 0, which the API already refuses on the same
 * column. The form judges nothing itself (ADR 0052): every complaint on screen
 * came from the server.
 */
function manualSaleQuantity(raw: string): number {
  const quantity = Number(raw.trim());
  return Number.isFinite(quantity) ? quantity : 0;
}

/**
 * manualSaleAmount reads the amount cell into cents, or null for a blank one.
 *
 * Blank is null and means the Ticket Type's catalog price. Zero is zero and
 * records a comp — collapsing it into blank would charge the catalog price for
 * a sale the organizer stated was free.
 *
 * The amount is PER TICKET, as it is in the uploaded template and in the Sale
 * Correction, and the label says so.
 */
function manualSaleAmount(raw: string): number | null {
  if (raw.trim() === "") {
    return null;
  }
  const price = Number(raw);
  return Number.isFinite(price) ? Math.round(price * 100) : null;
}

/**
 * manualSaleBody is the ONE reading of the form the API sees. The live preview
 * and the save send exactly this, so the verdict shown is the verdict recorded
 * on — the same guarantee the Sale Correction's form makes (#352).
 *
 * The zone is the Event's: a sale date typed as a wall-clock moment is only an
 * instant once somebody says whose clock it was on.
 */
export function manualSaleBody(form: ManualSaleForm, zone: string): RecordSaleInput {
  return {
    customer_email: form.customerEmail.trim(),
    customer_first_name: form.customerFirstName.trim(),
    customer_last_name: form.customerLastName.trim(),
    customer_tax_id_type: form.taxIdType,
    customer_tax_id_number: form.taxIdNumber.trim(),
    ticket_type_id: form.ticketTypeId,
    quantity: manualSaleQuantity(form.quantity),
    payment_method: form.paymentMethod,
    sold_at: dateTimeLocalToISO(form.soldAtLocal, zone) ?? "",
    amount_cents: manualSaleAmount(form.unitPrice),
  };
}

/** What the form shows of a preview verdict, and whether it may be saved. */
export type ManualSaleVerdict = {
  // True when the save would be refused: the row is invalid, the Ticket Type is
  // oversold or over the Purchase Limit, or the preview judged no row at all.
  blocks: boolean;
  // The row's complaints by template column, first complaint per column.
  fieldErrors: Record<string, string>;
  // The sold-at date of another active sale this one matches, or null. A
  // warning: it never blocks.
  duplicateOfDate: string | null;
  // What the chosen Ticket Type has left once this sale is recorded, or null
  // when the preview named no Ticket Type.
  remaining: { ticketTypeName: string; remaining: number } | null;
};

/**
 * manualSaleVerdict reads a preview into what the form needs to decide.
 *
 * Stated here rather than borrowed from the Sale Correction's reading of the
 * same shape, because the two answer different questions with it: a correction's
 * capacity is netted against the sale being reversed, and this one's is the
 * plain remainder on an Event nothing is being given back to. The arithmetic
 * agreeing today is a coincidence of that netting being zero here.
 *
 * The duplicate never blocks. Everything the API would refuse does.
 */
export function manualSaleVerdict(result: ImportPreviewResult): ManualSaleVerdict {
  const row = result.rows[0];
  if (!row) {
    return { blocks: true, fieldErrors: {}, duplicateOfDate: null, remaining: null };
  }
  const fieldErrors: Record<string, string> = {};
  for (const complaint of row.errors ?? []) {
    if (!(complaint.field in fieldErrors)) {
      fieldErrors[complaint.field] = complaint.message;
    }
  }
  const impact = result.capacity_impact[0];
  return {
    blocks: !row.valid || !result.committable,
    fieldErrors,
    duplicateOfDate: row.possible_duplicate && row.duplicate_of_date ? row.duplicate_of_date : null,
    remaining: impact
      ? { ticketTypeName: impact.ticket_type_name, remaining: impact.remaining - impact.requested }
      : null,
  };
}

/**
 * The form to hand back after a sale is recorded and the sitting continues.
 *
 * The keep-adding session itself is #372 and is deliberately not built here.
 * This rule is, because it is a fact about THIS form's state — which of its
 * cells belong to the buyer and which to the sitting — and #372 wraps the form
 * rather than reaching inside it.
 *
 * The buyer, their Tax ID and the amount clear; the Ticket Type, Payment Method
 * and sale date stay, because twenty door sales from the same night are the same
 * three answers twenty times. The quantity returns to one and the amount clears
 * rather than persisting: a stale amount is the one sticky value that would
 * record the wrong money while looking correct.
 */
export function manualSaleFormAfterSave(form: ManualSaleForm): ManualSaleForm {
  return {
    ...emptyManualSaleForm(form.soldAtLocal),
    ticketTypeId: form.ticketTypeId,
    paymentMethod: form.paymentMethod,
  };
}
